crates/autopilot/src/database/order_events.rs

```diff
@@ -16,21 +17,65 @@ impl super::Postgres {
     /// elaborate error handling is not necessary because this is just
     /// debugging information.
     pub async fn store_order_events(&self, events: &[(OrderUid, OrderEventLabel)]) {
-        if let Err(err) = store_order_events(self, events, Utc::now()).await {
+        if let Err(err) = store_all_order_events(self, events, Utc::now()).await {
             tracing::warn!(?err, "failed to insert order events");
         }
     }
 
+    /// Inserts the given events with the current timestamp into the DB for the
+    /// market orders only.
+    pub async fn store_market_order_events(&self, events: &[(OrderUid, OrderEventLabel)]) {
+        if let Err(err) = store_market_order_events(self, events, Utc::now()).await {
+            tracing::warn!(?err, "failed to insert market order events");
+        }
+    }
+
     /// Deletes events before the provided timestamp.
     pub async fn delete_order_events_before(&self, timestamp: DateTime<Utc>) -> Result<u64, Error> {
         order_events::delete_order_events_before(&self.0, timestamp).await
     }
 }
 
+#[async_trait]
+pub trait OrderEventPersister {
+    async fn insert_order_event(
+        &self,
+        conn: &mut PgConnection,
+        event: &OrderEvent,
+    ) -> Result<(), Error>;
+}
+
+struct RegularOrderEventPersister;
+
+#[async_trait]
+impl OrderEventPersister for RegularOrderEventPersister {
+    async fn insert_order_event(
+        &self,
+        conn: &mut PgConnection,
+        event: &OrderEvent,
+    ) -> Result<(), Error> {
+        order_events::insert_order_event(conn, event).await
+    }
+}
+
+struct MarketOrderEventPersister;
+
+#[async_trait]
+impl OrderEventPersister for MarketOrderEventPersister {
+    async fn insert_order_event(
+        &self,
+        conn: &mut PgConnection,
+        event: &OrderEvent,
+    ) -> Result<(), Error> {
+        order_events::insert_market_order_event(conn, event).await
+    }
+}
+
 async fn store_order_events(
     db: &super::Postgres,
     events: &[(OrderUid, OrderEventLabel)],
     timestamp: DateTime<Utc>,
+    persister: &impl OrderEventPersister,
```

Comment:
Should be able to pass a Fn here instead of an entire trait.

Reply:
Already discussed [here](https://github.com/cowprotocol/services/pull/2121#discussion_r1415533987). BTW I tried my hand on it and it's a PITA to get it right. The issue is that we have to pass a closure that itself takes references and making the compiler understand the lifetimes of each reference is nightmarish. 😱

Reply:
BTW even if we can't implement the "optimal" solution without jumping through several hoops I think simply duplicating the code is probably still better than introducing the trait just so we can shared a tiny bit of trivial code.